src/metabase/models/database.clj

```diff
@@ -69,10 +71,60 @@
+  each iteration step, if there is a -value suffixed property set in the details to be persisted, then we instead insert
+  (or update an existing) Secret instance and point to the inserted -id instead."
+  [database details conn-prop-nm conn-prop]
+  (let [sub-prop  (fn [suffix]
+                    (keyword (str conn-prop-nm suffix)))
+        id-kw     (sub-prop "-id")
+        new-name  (format "%s for %s" (:display-name conn-prop) (:name database))
+        ;; in the future, when secret values can simply be changed by passing
+        ;; in a new ID (as opposed to a new value), this behavior will change,
+        ;; but for now, we should simply look for the value
+        value-kw  (sub-prop "-value")
+        value     (value-kw details)
+        kind      (:secret-kind conn-prop)
+        source-kw (sub-prop "-source")
+        source    (source-kw details)]
+    (if (nil? value) ;; secret value for this conn prop was not changed
+      details
+      (let [{:keys [id creator_id created_at]} (secret/upsert-secret-value!
+                                                 (id-kw details)
+                                                 new-name
+                                                 kind
+                                                 source
+                                                 value)]
+        ;; remove the -value keyword (since in the persisted details blob, we only ever want to store the -id)
+        (-> details
+          (dissoc value-kw)
+          (assoc id-kw id)
+          (assoc (sub-prop "-creator-id") creator_id)
+          (assoc (sub-prop "-created-at") (t/format :iso-offset-date-time created_at)))))))
+
+(defn- handle-secrets-changes [{:keys [details] :as database}]
+  (let [conn-props-fn (get-method driver/connection-properties (driver.u/database->driver database))]
+    (cond (nil? conn-props-fn)
+          database ; no connection-properties multimethod defined; can't check secret types so do nothing
+
+          details ; we have details populated in this Database instance, so handle them
+          (let [conn-props            (conn-props-fn (driver.u/database->driver database))
+                conn-secrets-by-name  (->> (filter #(= "secret" (:type %)) conn-props)
+                                           (reduce (fn [acc prop] (assoc acc (:name prop) prop)) {}))
+                ;; saved-details         (db/select-field :details Database id)
+                 updated-details      (reduce-kv (partial handle-db-details-secret-prop! database)
+                                                 details
+                                                 conn-secrets-by-name)]
+           (assoc database :details updated-details))
+
+          :else ; no details populated; do nothing
+          database)))
+
 ;; TODO - this logic would make more sense in post-update if such a method existed
```

Comment:
this comment should be moved down as the `(handle-secrets-changes database)` part does not belong in the hypothetical `post-update`

Reply:
OK, I think I see what you mean.  See update in latest commit.

Reply:
Yep. That's what I meant.

I'm not yet familiar enough with the code base to review the rest of the PRs so I won't be approving here.